core/src/mast/multi_forest_node_iterator.rs

```diff
@@ -0,0 +1,508 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+
+use miden_crypto::hash::rpo::RpoDigest;
+
+use crate::mast::{MastForest, MastForestError, MastNode, MastNodeId};
+
+type ForestIndex = usize;
+
+/// Depth First Search Iterator in Post Order for [`MastForest`]s.
+///
+/// This iterator iterates through all **reachable** nodes of all given forests exactly once.
+///
+/// Since a `MastForest` has multiple possible entrypoints in the form of its roots, a depth-first
+/// search must visit all of those roots and the trees they form. This iterator's `Item` is
+/// [`MultiMastForestIteratorItem`]. It contains either a [`MultiMastForestIteratorItem::Node`] of a
+/// forest, or the replacement of an external node. This is returned if one forest contains an
+/// External node with digest `foo` and another forest contains a non-external node with digest
+/// `foo`. In such a case the `foo` node is yielded first (unless it was already visited) and
+/// subsequently a "replacement signal" ([`MultiMastForestIteratorItem::ExternalNodeReplacement`])
+/// for the external node is yielded to make the caller aware that this replacement has happened.
+///
+/// All of this is useful to ensure that children are always processed before their parents, even if
+/// a child is an External node which is replaced by a node in another forest. This guarantees that
```

Comment:
I may have missed discussion on this elsewhere - but I must insist that we not rely on this property (that child nodes and parent nodes have a guaranteed relationship in terms of their `MastNodeIds`) anywhere - it means modifying the forest after it is constructed is all but impossible to do without violating this property. We may not have a reason to do so now, but doing so will be effectively dead on arrival if we come to rely on this, and I can imagine all sorts of reasons why we might _want_ to do that.

My understanding is that any reliance on this property is basically unintentional/out of convenience, not necessity, IIRC @plafer mentioned it is only the (de)serialization code? In any case, I think we should treat this as a bug to be fixed in _that_ code, rather than a reason to impose constraints on construction/merging of `MastForest`s.

Reply:
It also occurs to me that we can have multiple callers to the same procedure in different forests, so a naive postorder traversal will _not_ uphold this guarantee. You would need to first compute a post-dominator tree covering both forests, and use _that_ to ensure that you visit all roots in postorder as well.

Reply:
Just for context, this was discussed here, though more the how and not the why: https://github.com/0xPolygonMiden/miden-vm/pull/1534#discussion_r1807487205

Also for reference, yes the serialization code was (most likely) the only place and it was addressed here https://github.com/0xPolygonMiden/miden-vm/pull/1542, although I was about to close that PR since we have a solution now, but if you consider it a "bug", we can also consider merging it.

It's important to note that we do not _rely_ on the property, we just establish it during merging. So even if we need to violate this property in the future, I don't think anything here would break. Correct me if I'm wrong, but even if we construct a tree like `[Join(1,2), Block(foo), Block(bar)]` then doing a DFS traversal would always cleanly yield an order where "child id < parent id". I think this would only break if there are loops in the graph, but that would be a problem for other reasons as well.

So unless it is a problem that we create this order (which I think it isn't, since the assembler or MastForestBuilder does that too?) I think we're fine.

Reply:
The issue isn't that we happen to incidentally create a forest with this property, that's obviously fine - but more that we might end up relying on it, unintentionally or otherwise.

> Correct me if I'm wrong, but even if we construct a tree like [Join(1,2), Block(foo), Block(bar)] then doing a DFS traversal would always cleanly yield an order where "child id < parent id". I think this would only break if ther

Because we visit roots in effectively random order, we can't guarantee that the order in which we visit nodes is topologically sorted, and thus the order in which we assign node ides isn't guaranteed to uphold the "children before parents" property when children can have multiple parents from different roots.

The only way to uphold that guarantee, would be to visit all nodes in all forests every time they are referenced (which is only true because we don't need to consider dynamic invocation, and so the graph has no cycles), meaning we will enqueue and visit the same nodes (and their children!) `O(callers)` number of times even though they have already been visited. Will it work? Yes, but at the cost of exponential complexity hiding in the algorithm. If you end up merging a forests where many callers call a large procedure (or if there are multiple such "hot spots" in the call graph)...as you can imagine, that could get out of hand quick.

Reply:
To add on to the above, here's how you'd avoid the need for the exponential solution:

1. Perform a depth-first visit of all unique roots in all forests, visiting each root only once. While visiting a root, when you reach a reference to another procedure root, add it to the set of successors for the current root, ensuring each root is in the successor set only once, and finish visiting the rest of the procedure (if there remains anything to be visited in the same procedure) until all successors have been seen. Add edges from the current root to all of its successors in an adjacency matrix-style graph (i.e. `BTreeMap<Digest, Vec<Digest>>`). Then, add each successor to the worklist (if it hasn't already been visited), and start visiting the next root in the worklist. Repeat until the worklist is empty. The resulting graph represents a complete dependency graph for procedures in the merged forest.
2. Compute the topological ordering of the graph computed in 1. Essentially, you add all roots of the graph to a worklist of nodes, and start visiting them. When you visit a node, and it has successors in the dependency graph, you add the root back to the worklist, then pop one of its successors from the dependency graph and add it to the worklist. If a node has no successors remaining the dependency graph, you add it to the vector representing the topological order. When the worklist is empty, you've topologically ordered the graph. The order of nodes in the topological order will be ordered from leaves to roots (since nodes are added to the order once all their successors have been visited).
3. Copy nodes into the merged forest following the topological order of the procedure roots computed above.

This sounds expensive, but only requires visiting the entire forest twice: once to compute the dependency graph, and once to assign node ids and construct the merged forest. It will be an order of magnitude faster than visiting all subtrees of the forest N times, where N is the number of parents for a given subtree.

Reply:
My understanding is that this is what the code is kind of doing already: there are two passes over all nodes:

1. The first pass to build a map of all non-external nodes (the map is `node_digest -> (forest_idx, node_id)`).
2. The second pass uses the above map to make sure that when we encounter an external node which has an equivalent non-external node in any of the forests, we visit the tree corresponding to the non-external node first.

So, as far as I can tell, there is no exponential blowup, but @PhilippGackstatter can correct me if I'm wrong.

Reply:
> Because we visit roots in effectively random order, we can't guarantee that the order in which we visit nodes is topologically sorted, and thus the order in which we assign node ids isn't guaranteed to uphold the "children before parents" property when children can have multiple parents from different roots.

I don't understand why it matters in which order we visit _roots_, if we're doing a depth-first traversal.

For instance, consider a Mastforest `X` like this with all of the shown nodes being roots, for example's sake:

```text
  A
 / \
B   C
 \ /
  D
```

- It does not matter at which of the roots we start: We will always add `D` to the merged forest first due to DFS. Whether `B` or `C` is added next depends where we start but in any case both are valid topogical orderings. And `A` will always be added last. So we get either one of these orderings `[D, B, C, A]` or `[D, C, B, A]`, both of which are fine.

Dealing with multiple forests isn't changing much, only External nodes and duplicates are "special" cases. To illustrate, perhaps we have another Mast Forest `Y` which has an External node whose referenced digest matches that of a node in the previous forest:

```text
  E
 /
Ext(A)
```

- If we process this graph next, then `Ext(A)` will be detected as a duplicate of `A` in the merged forest and remapped. `E` will be added on top and the property holds.
- If we were to first merge `Y` and then `X`, the iterator would descend down from `E` and first traverse `A` instead of `Ext(A)` (and later remap the external node to the new index of `A`) and the node id property would also be upheld for the entire merged forest, since the graph of `A` will be processed in its entirety before `E`.

Which is all to say that I don't think it matters in which order we process the _roots_ due to the depth-first traversal. The depth-first traversal creates a topological ordering.

> The only way to uphold that guarantee, would be to visit all nodes in all forests every time they are referenced

Indeed this is what the code did at the time of your writing and I agree that is inefficient. This was necessary for reasons explained [here](https://github.com/0xPolygonMiden/miden-vm/blob/310f90341c3b9128d3527c9d300d66dbf620f660/core/src/mast/multi_forest_node_iterator.rs#L463-L488) but it's no longer necessary now.

Fortunately it is possible to make a slight modification to use a `VecDeque` instead of a stack (`Vec`), and so we can discover _and_ process nodes in postorder (whereas previously they were discovered in preorder but processed in postorder, which caused the issues). This has the effect that the first time we visit a node it will be marked as visited and not processed again later which makes the implementation no longer have exponential complexity afaict.
So for forest `X`, the iterator would build this deque: `[D, B, C, A]` and by popping elements off the front we visit nodes in postorder and therefore the children before the parents.

If you think this is not correct yet, can you give an example of a MastForest where this wouldn't work?

Reply:
> Which is all to say that I don't think it matters in which order we process the _roots_ due to the depth-first traversal. The depth-first traversal creates a topological ordering.

I think perhaps what was not clear/apparent to me, is at what point the node identifiers are being assigned. When you can visit the entire forest in a single traversal, and you assign node ids by popping nodes off the visit stack not on the way back up the tree, but _after_ visiting the entire tree, then you don't run into problems, since the visit stack will necessarily be a valid topological order. 

If you were instead to assign them on the way back up, then the order would be trivially easy to invalidate:

```mermaid
flowchart TD
    A
    B
    C
    D
    E
    F
    A --> B
    B --> C
    C --> D
    B --> F
    F --> E
    C --> F
```

Here, if you visit `A -> B -> F -> E`, and assign node ids on the way back up, i.e. `E, F`  _then_ start descending `C -> D`, and assign on the way back up, i.e. `D -> C -> B -> A`, then `F` has a node id greater than `C`. That is a valid postorder DFS traversal.

So if you instead wait until visiting the entire tree above, then your visit stack is `E, F, D, C, B, A`, and popping nodes off that stack lets you guarantee that parents are assigned ids before children (and I _think_ the inverse, if you were to just iterate over the stack like a vector, but I'm not sure if that's a guaranteed property, or incidental - I haven't been able to contrive a counter-example yet though, so).

I think that's likely where I misconstrued how this was implemented (since it was described as assigning node ids in postorder, which is _true_, but also incomplete, node ids are actually being assigned in post-dominance order, assuming the approach I described in my previous paragraph.

I'm not sure if that also extends to assigning ids after each traversal from a single procedure root (and it's reachable children). I _think_ it works out the same way.

> Indeed this is what the code did at the time of your writing and I agree that is inefficient. This was necessary for reasons explained [here](https://github.com/0xPolygonMiden/miden-vm/blob/310f90341c3b9128d3527c9d300d66dbf620f660/core/src/mast/multi_forest_node_iterator.rs#L463-L488) but it's no longer necessary now.

Ah, I think I saw a note in the code about this, which is what prompted my note about exponential complexity.

> Fortunately it is possible to make a slight modification to use a `VecDeque` instead of a stack (`Vec`), and so we can discover _and_ process nodes in postorder (whereas previously they were discovered in preorder but processed in postorder, which caused the issues). This has the effect that the first time we visit a node it will be marked as visited and not processed again later which makes the implementation no longer have exponential complexity afaict. So for forest `X`, the iterator would build this deque: `[D, B, C, A]` and by popping elements off the front we visit nodes in postorder and therefore the children before the parents.

I'll have to re-read the code to better understand this - AFAIK, graph theory requires a preorder visit to construct the postorder, so it isn't apparent to me how a deque changes that. Unless what you mean is that you still perform the preorder traversal, but how you consume the visit stack uses the deque (basically what I mentioned above about constructing the visit stack, and then consuming from the front to assign ids). If that's the case, I don't know if you even need the deque, you can just consume the vector as an iterator, but that's two sides of the same stone. In any case, if that's what you mean, that should work just fine.

> If you think this is not correct yet, can you give an example of a MastForest where this wouldn't work?

I think after exploring things a bit more, I think the way ids are assigned during the merge upholds the "children before parents" property, at least, I'm unable to find a counter-example, and working through it formally, I think it is probably guaranteed. I'm still opposed to making that a public part of the contract of MastForest, I would even erase any mention of it as an incidental property in the docs. It will make any of the following harder/impossible without significant rework, and could end up being relied on in various places (just like the fact that the MastForestBuilder constructs forests with that property, ended up being relied on incidentally):

* Mutating forests after construction/merging, so as to perform optimizations (which would then require recomputing all of the affected hashes, but the node ids would not need to change if we aren't imposing this children have older ids than parents property). In this case, you might remove some edges or nodes, or both, and insert new nodes/edges, and new nodes would almost certainly end up having younger indexes than their parents.
* Supporting recursion in some form other than `DYN`, not something we need to really worry about right now, but I think it is very likely that research in this area could open up the opportunity to support it in some fashion, and it would be best if our program representation isn't overly dependent on the fact that cycles can't happen. `DYN` already makes it possible for cycles to _exist_, they just don't manifest in a way that is a problem currently.

Anyway, I think we can probably resolve this for now. If an issue crops up with a real program, we can revisit things at that point, but I don't think we need to hold this PR up.

Reply:
> I'm still opposed to making that a public part of the contract of MastForest

I agree with this. Even though I've asked to maintain this property during merging, this should not be a guarantee that we provide in the future. So, we should go ahead finish https://github.com/0xPolygonMiden/miden-vm/pull/1542 to make sure this is not assumed during deserialization.

I would merge this PR as is however and create a separate issue to update docs to make it clear that `MastForest` does not provide this guarantee.

Reply:
> AFAIK, graph theory requires a preorder visit to construct the postorder

I'm not sure I understand this point or that that is the case.
I'm traversing the graph in a postorder manner, e.g. snippets from the code:

https://github.com/0xPolygonMiden/miden-vm/blob/0f76408b3d42faa7ee8b5baa9a7dbd998db66c94/core/src/mast/multi_forest_node_iterator.rs#L158-L166

The first child's tree is visited before the second child's which is visited before the node itself (postorder). As the nodes are being visited they are added to the deque (at the back) and later popped off the front, yielding the same order as the one they were added in to the caller. Since it yields a postorder I think it's correct, but perhaps we still mean something different.

> you can just consume the vector as an iterator, but that's two sides of the same stone

Yes, that or we could use a `Vec` and reverse it after a subgraph has been discovered and pop elements off the stack (which wouldn't need any additional housekeeping like storing the index of the last visited node. Note that the iterator as implemented discovers the graph lazily. Only after a subgraph has been completely visited is the next subgraph discovered, and so on. If we were constructing the graph eagerly, then consuming the `Vec` as an iterator would be the simplest I agree).

Overall, the `VecDeque` usage makes the code easier to follow and understand than using a `Vec` since no code operations have to be reversed and the stack does not need to be reversed. Since the deque has just an 8 bytes higher memory footprint I think that's fine so, I'm happy to move forward with the PR as-is.

Reply:
> I'm not sure I understand this point or that that is the case. I'm traversing the graph in a postorder manner, e.g. snippets from the code:

That snippet demonstrates a _pre-order_ traversal, _post-order_ visit. In other words, you must traverse (and visit) the entire subgraph rooted under a node, before you "visit" that node. The terminology can be confusing because _visit_ can mean two different things depending on whether you are talking about taking action on a node when it is _first discovered_ (pre-order), or when all of its successors have been visited (post-order).

Any post-order visitor of a graph can _also_ be a pre-order visitor by simply performing an action upon discovering a node in addition to the post-order action when all of it's successors are done being visited.

So my point was, a deque as the output of a post-order visitor doesn't actually gain you anything, so far as I can tell. To be clear, not saying the use of one is _bad_ or incorrect (unless I'm missing something), it just wasn't clear to me how using one would differ _in practice_ from forward-iterating a vector.

> Note that the iterator as implemented discovers the graph lazily
>
> Only after a subgraph has been completely visited is the next subgraph discovered, and so on. If we were constructing the graph eagerly, then consuming the Vec as an iterator would be the simplest I agree).

If we're consuming nodes from the iterator as they are produced, then you don't actually need a separate vector/deque for storing the results of the iterator at all - just the stack used internally to manage the traversal. If we actually want to materialize the full set in a collection, then I'd assume that a vector would be sufficient.

In any case, I think we're talking past each other here - the output is correct currently, so I think I'm probably just being pedantic about terminology. On the off-chance that discussing this surfaces something we didn't hit during review, I wanted to make sure I clarified things, but AFAIK we're good.

Reply:
My understanding here is that using `VecDeque ` is fine, but we could also just use regular `Vec` (and maybe that would be a bit simpler) like so:

1. During discovery, push new nodes to the end of the vector.
2. When returning the nodes from the node iterator, just use the regular vector iterator (this would be identical to removing nodes from the front of `VecDeque`).
3. Once the iterator is exhausted, truncate the vector and start the discovery again as in step 1.

But, I think this is a relatively small potential improvement over the existing approach - so, I'd have a separate issue for it.